import subprocess
import os
from pathlib import Path

import polars as pl

def _plink_merge_bim_files(bim_file_main: Path, bim_files_to_merge: list[Path], output_bim_file: Path,
                            plink_path: Path =Path("plink/plink")) -> None:

    """
    Merges two bim files using PLINK.

    Args:
        bim_file_main (str): The path to the main BIM file (without extension).
        bim_file_to_merge (str): The path to the file to merge (without extension).
        plink_path (str): The path to the PLINK executable (default: "plink/plink").

    Returns:
        str: The path to the merged BIM file (without extension).
    """

    bim_list = os.path.join(bim_file_main, "_list_to_merge.txt")
    with open(bim_list, 'w') as f:
        for bim_file in bim_files_to_merge:
            f.write(f"{bim_file}\n")

    merge_command = [
        plink_path,
        "--dog",
        "--bfile", bim_file_main,
        "--merge-list", bim_list,
        "--out", output_bim_file
    ]

    try:
        # Execute the PLINK command
        result = subprocess.run(merge_command, check=True, text=True, capture_output=True)
        print("PLINK output:", result.stdout)
    except subprocess.CalledProcessError as e:
        print("Error running PLINK:", e.stderr)
        raise
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        raise

def _plink_convert_tped_to_bim(tped_file: Path, output_file: Path, plink_path: Path=Path("plink/plink")) -> None:
    """
    Converts a TPED file to bim format using PLINK.

    Args:
        tped_file (str): The path to the TPED file (without extension).
        plink_path (str): The path to the PLINK executable.
        output_file (str): The desired output BIM file path (without extension).
    """

    convert_command = [
        plink_path,
        "--dog",
        "--tfile", tped_file,
        "--make-bed",
        "--out", output_file
    ]

    try:
        # Execute the PLINK command
        result = subprocess.run(convert_command, check=True, text=True, capture_output=True)
        print("PLINK output:", result.stdout)
    except subprocess.CalledProcessError as e:
        print("Error running PLINK:", e.stderr)
        raise
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        raise

def _plink_produce_genome_file(tped_file_main: Path, tped_files_to_merge: list[Path],
                            output_genome_file: Path =Path(""), plink_path: Path =Path("plink/plink")) -> None:
    if output_genome_file == "":
        output_genome_file = Path("ibd/" + os.path.basename(tped_file_main))

    merged_bim_file = Path(os.path.join(tped_file_main, "_merged"))

    # Converts tped files to bim files
    _plink_convert_tped_to_bim(
        tped_file=tped_file_main,
        output_file=tped_file_main,
        plink_path=plink_path
    )
    for tped_file in tped_files_to_merge:
        _plink_convert_tped_to_bim(
            tped_file=tped_file,
            output_file=tped_file,
            plink_path=plink_path
        )

    # Merges parents' bim files with offspring bim file
    _plink_merge_bim_files(
        bim_file_main=tped_file_main,
        bim_files_to_merge=tped_files_to_merge,
        plink_path=plink_path,
        output_bim_file=merged_bim_file
    )

    # Produces .genome file
    roh_command = [
        plink_path,
        "--dog",
        "--bfile", merged_bim_file,
        "--genome",
        "--out", output_genome_file
    ]

    try:
        # Execute the PLINK command
        result = subprocess.run(roh_command, check=True, text=True, capture_output=True)
        print("PLINK output:", result.stdout)
    except subprocess.CalledProcessError as e:
        print("Error running PLINK:", e.stderr)
        raise
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        raise

def plink_roh(input_file: Path, output_folder: Path =Path("roh/"), plink_path: Path =Path("plink/plink"),
                window_snp: int = 50, window_het: int = 1, window_missing: int = 5,
                window_threshold: float = 0.05, homozyg_gap: int = 1000, homozyg_het: int = 1000,
                homozyg_density: int = 50, homozyg_snp: int = 100, homozyg_kb: int = 1000) -> None:

    """
    Sends a file to PLINK for Run of Homozygosity (RoH) analysis.

    Args:
        Filesystem:
            input_file (str):               Path to the input file for PLINK, without extension
            output_prefix (str):            Prefix for the output files generated by PLINK
            plink_path (str):               Path to the PLINK executable (default: "plink")
        PLINK Parameters (with default values):
            Defining the scanning window:
                window_snp (int):           Size of the scanning window in SNPs (default: 50)
                window_het (int):           Maximum number of heterozygous calls allowed in the scanning window (default: 1)
                window_missing (int):       Maximum number of missing calls allowed in the scanning window (default: 5)
                window_threshold (float):   Minimum scanning window hit rate (default: 0.05)
            Tresholds for calling a homozygous segment:
                homozyg_gap (int):          Maximum interval between two SNPs in a segment in kilobases (default: 1000)
                homozyg_het (int):          Maximum number of heterozygous calls in a homozygous segment (default: 1000) - This flag is disabled by default in PLINK
                homozyg_density (int):      Maximum inverse density of a homozygous segment (default: 50)
                homozyg_snp (int):          Minimum number of SNPs in a homozygous segment (default: 100)
                homozyg_kb (int):           Minimum length of a homozygous segment in kilobases (default: 1000)

    Outputs:
        Generates output files in the specified output folder:
            .hh         	                List of het. haploid and nonmale Ychr genotypes.
            .hom	                        Run-of-homozygosity list.
            .hom.indiv		                Sample-based runs-of-homozygosity report.
            .hom.summary    	            SNP-based runs-of-homozygosity report.
    """

    if not os.path.exists(os.path.join(input_file, ".tped")):
        raise FileNotFoundError(f"Input file '{input_file}' does not exist.")

    # Construct the PLINK command
    roh_command = [
        plink_path,
        "--tfile", input_file,
        "--dog",
        "--homozyg",
        "--homozyg-window-snp", str(window_snp),
        "--homozyg-window-het", str(window_het),
        "--homozyg-window-missing", str(window_missing),
        "--homozyg-window-threshold", str(window_threshold),
        "--homozyg-gap", str(homozyg_gap),
        "--homozyg-het", str(homozyg_het),
        "--homozyg-density", str(homozyg_density),
        "--homozyg-snp", str(homozyg_snp),
        "--homozyg-kb", str(homozyg_kb),
        "--out", output_folder
    ]

    try:
        # Execute the PLINK command
        result = subprocess.run(roh_command, check=True, text=True, capture_output=True)
        print("PLINK output:", result.stdout)
    except subprocess.CalledProcessError as e:
        print("Error running PLINK:", e.stderr)
        raise
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        raise

def plink_parentage(offspring_file: Path, parent1_file: Path, parent2_file: Path, genome_file: Path,
                    plink_path: Path =Path("plink/plink")) -> None:
    """
    Sends the target individual and potential parents' files to PLINK for parentage analysis.

    Args:
        offspring_file (Path): The path to the file containing the offspring's genotypes.
        parent1_file (Path): The path to the file containing the first parent's genotypes.
        parent2_file (Path): The path to the file containing the second parent's genotypes.
        output_bim_file (Path): The desired output BIM file path (without extension). If None, defaults to offspring_file + "_merged".
        plink_path (Path): The path to the PLINK executable (default: "plink/plink").
    """

    if not os.path.exists(os.path.join(offspring_file, ".tped")):
        raise FileNotFoundError(f"Offspring file '{offspring_file}' does not exist.")
    if not os.path.exists(os.path.join(parent1_file, ".tped")):
        raise FileNotFoundError(f"Parent 1 file '{parent1_file}' does not exist.")
    if not os.path.exists(os.path.join(parent2_file, ".tped")):
        raise FileNotFoundError(f"Parent 2 file '{parent2_file}' does not exist.")

    _plink_produce_genome_file(
        tped_file_main=offspring_file,
        tped_files_to_merge=[parent1_file, parent2_file],
        output_genome_file=genome_file,
        plink_path=plink_path
    )

    genome = pl.read_csv(os.path.join(genome_file, ".genome"), separator="\t")



